### PR TITLE
Add Schnorrsig half aggregation

### PR DESCRIPTION
**Pull request by jonasnick**
Result of a recorded pair-programming session with @real-or-random to demonstrate how to write libsecp(-zkp) code. Therefore the code is still in a very early stage.

### CODE REVIEW

**Comment by jonasnick** [comment]
This doesn't allow aggregating aggregate sigs. Would be nice if it could.

File: src/modules/schnorrsig/main_impl.h [hunk 1]
```diff
@@ -236,4 +236,163 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
...
+    secp256k1_scalar s;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(aggsig != NULL);
+    ARG_CHECK(aggsig_size != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    if (*aggsig_size < 32*(1 + n_sigs)) {
```
Comment (rustyrussell) [inline src/modules/schnorrsig/main_impl.h:282]: Overflow protection needed here, AFAICT.  You probably want to divide aggsig_size, check it's > 0, then sub 1 and compare with n_sigs.

File: include/secp256k1_schnorrsig.h [hunk 2]
```diff
@@ -104,6 +106,25 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_schnorrsig_verify(
...
 ) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4);
 
+SECP256K1_API int secp256k1_schnorrsig_aggregate(
+    const secp256k1_context* ctx,
+    unsigned char* aggsig,
+    size_t* aggsig_size,
+    unsigned char **sig64,
+    unsigned char **msg32,
+    secp256k1_xonly_pubkey *pubkey,
+    uint32_t n_sigs
```
Comment (rustyrussell) [inline include/secp256k1_schnorrsig.h:116]: I know C is sloppy with this, but const unsigned char **sig64 would be nice.  And msg32.

Also, n_sigs here and n_msgs below?

File: include/secp256k1_schnorrsig.h [hunk 3]
```diff
@@ -104,6 +106,25 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_schnorrsig_verify(
...
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4) SECP256K1_ARG_NONNULL(5) SECP256K1_ARG_NONNULL(6);
+
+SECP256K1_API int secp256k1_schnorrsig_aggverify(
+    const secp256k1_context* ctx,
+    unsigned char **msg32,
+    uint32_t n_msgs,
+    secp256k1_xonly_pubkey *pubkey,
```
Comment (rustyrussell) [inline include/secp256k1_schnorrsig.h:123]: const here too?

File: include/secp256k1_schnorrsig.h (hunk 2 above)
Comment (rustyrussell) [reply include/secp256k1_schnorrsig.h:116]: And sig64 adn msg32 are arrays of ptrys, pubkey is a direct array of objects.  That's a bit weird?  Also, please `pubkeys` to give the poor user a hint?

File: src/modules/schnorrsig/main_impl.h [hunk 4]
```diff
@@ -236,4 +236,163 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
...
+    secp256k1_scalar_set_int(&s, 0);
+    for (i = 0; i < n_sigs; i++) {
+        unsigned char randomizer[32];
+        secp256k1_scalar zi;
+        secp256k1_scalar si;
+
+        secp256k1_sha256_initialize(&hash);
+        secp256k1_sha256_write(&hash, midhash, sizeof(midhash));
+        /* TODO: fix endianness issue */
+        secp256k1_sha256_write(&hash, (unsigned char *)&i, sizeof(i));
```
Comment (rustyrussell) [inline src/modules/schnorrsig/main_impl.h:300]: Yes.  Please do :)

File: src/modules/schnorrsig/main_impl.h [hunk 5]
```diff
@@ -236,4 +236,163 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
...
+    secp256k1_scalar s;
+    int overflow;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(aggsig != NULL);
+
+    secp256k1_gej_set_infinity(&rhs);
+    if (aggsig_size != 32*(1 + n_msgs)) {
```
Comment (rustyrussell) [inline src/modules/schnorrsig/main_impl.h:331]: Overflow

File: src/modules/schnorrsig/main_impl.h [hunk 6]
```diff
@@ -236,4 +236,163 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
...
+        secp256k1_fe rx;
+        secp256k1_ge rp, pp;
+        secp256k1_scalar ei;
+        unsigned char pk_ser[32];
+        secp256k1_gej ppj, acc;
+
+        secp256k1_sha256_initialize(&hash);
+        secp256k1_sha256_write(&hash, midhash, sizeof(midhash));
+        /* TODO: fix endianness issue */
+        secp256k1_sha256_write(&hash, (unsigned char *)&i, sizeof(i));
```
Comment (rustyrussell) [inline src/modules/schnorrsig/main_impl.h:352]: Noted.

**Review by rustyrussell — COMMENTED** [review]
Rough review since I was reading it anyway.  Obv as PR states, needs more work!

**Comment by jonasnick** [comment]
We should make this PR compliant with the [draft spec](https://github.com/BlockstreamResearch/cross-input-aggregation/blob/master/half-aggregation.mediawiki) (which was written after this PR was opened).

**Comment by jonasnick** [comment]
Closing in favor of #268

**Pull request closed without merging**